### PR TITLE
feat: legalize onnx conv2d transpose to xtennn

### PR DESCRIPTION
**Pull request by ehsan-toosi**
- This PR provides xten-nn-to-torch legalization for conv2d transpose operation. The xtenn one doesn't have **output_shape** attribute anymore (like the one for onnx) and it's close to the definition of this operation for torch.aten. Therefore, it only has **output_padding** attribute for padding on the output. Everything is controlled via `output_padding`.
- xten-nn-to-torch also now supports dynamic shape and it was previously crashing on dynamic shapes.

### CODE REVIEW

File: lib/Conversion/XTenNNToTorch.cpp [hunk 1]
```diff
@@ -163,6 +169,37 @@ std::optional<ValueRange> groupConv2dToTorch(GroupConv2dOp op, GroupConv2dOp::Ad
...
+
+  auto stride = Torch::toTorchList(loc, rewriter, adaptor.getStride().vec());
+  auto dilation =
+      Torch::toTorchList(loc, rewriter, adaptor.getDilation().vec());
+  auto group =
+      rewriter.create<Torch::ConstantIntOp>(loc, adaptor.getGroupAttr());
+  auto outputPadding =
+      Torch::toTorchList(loc, rewriter, adaptor.getOutputPadding().vec());
+
+  // TODO: check the order is matched with what torch needs.
```
Comment (ttjost) [inline lib/Conversion/XTenNNToTorch.cpp]: Does it? 😄

File: test/Conversion/XTenNNToTorch/conv_transpose_2d.mlir [hunk 2]
```diff
@@ -0,0 +1,230 @@
+// RUN: aten-opt --convert-xtennn-to-torch  -split-input-file %s | FileCheck %s
+
+func.func @with_output_shape(%arg0: tensor<1x480x36x22xf32>, %arg1: tensor<480x360x4x4xf32>, %arg2: tensor<360xf32>) -> tensor<1x360x72x44xf32> {
+  %0 = xten_nn.ConvTranspose %arg0, %arg1, %arg2 {dilation = array<i64: 1, 1>, group = 1 : i64, output_padding = array<i64: 0, 0>, pad = [array<i64: 1, 1>, array<i64: 1, 1>], stride = array<i64: 2, 2>} : (tensor<1x480x36x22xf32>, tensor<480x360x4x4xf32>, tensor<360xf32>) -> tensor<1x360x72x44xf32>
```
Comment (ttjost) [inline test/Conversion/XTenNNToTorch/conv_transpose_2d.mlir:4]: Could you add a test where `group > 1` for completeness?

File: lib/Conversion/XTenNNToTorch.cpp (hunk 1 above)
Comment (roberteg16) [reply lib/Conversion/XTenNNToTorch.cpp]: I believe it only supports symmetric padding since only two inputs are expected. Can you confirm @ehsan-toosi ?

File: test/Conversion/XTenNNToTorch/conv_transpose_2d.mlir [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+// CHECK-NEXT:          %[[VAL_21:.*]] = torch.aten.conv_transpose2d.input %[[VAL_3]], %[[VAL_4]], %[[VAL_5]], %[[VAL_8]], %[[VAL_20]], %[[VAL_15]], %[[VAL_12]], %[[VAL_11]] : !torch.vtensor<[1,480,36,22],f32>, !torch.vtensor<[480,360,4,4],f32>, !torch.vtensor<[360],f32>, !torch.list<int>, !torch.list<int>, !torch.list<int>, !torch.int, !torch.list<int> -> !torch.vtensor<[1,360,75,47],f32>
+// CHECK-NEXT:          %[[VAL_22:.*]] = torch_c.to_builtin_tensor %[[VAL_21]] : !torch.vtensor<[1,360,75,47],f32> -> tensor<1x360x75x47xf32>
+// CHECK-NEXT:          return %[[VAL_22]] : tensor<1x360x75x47xf32>
+
+// -----
+
+func.func @with_pad_and_output_pad(%arg0: tensor<1x480x36x22xf32>, %arg1: tensor<480x360x4x4xf32>, %arg2: tensor<360xf32>) -> tensor<1x360x73x45xf32> {
+  %0 = xten_nn.ConvTranspose %arg0, %arg1, %arg2 {dilation = array<i64: 1, 1>, group = 1 : i64, output_padding = array<i64: 1, 1>, pad = [array<i64: 1, 1>, array<i64: 1, 1>], stride = array<i64: 2, 2>} : (tensor<1x480x36x22xf32>, tensor<480x360x4x4xf32>, tensor<360xf32>) -> tensor<1x360x73x45xf32>
+  return %0 : tensor<1x360x73x45xf32>
+}
```
Comment (roberteg16) [inline test/Conversion/XTenNNToTorch/conv_transpose_2d.mlir:99]: Can you use also asymmetric values for output padding and padding?

File: test/Conversion/XTenNNToTorch/conv_transpose_2d.mlir (hunk 2 above)
Comment (ehsan-toosi) [reply test/Conversion/XTenNNToTorch/conv_transpose_2d.mlir:4]: Done

File: test/Conversion/XTenNNToTorch/conv_transpose_2d.mlir (hunk 3 above)
Comment (ehsan-toosi) [reply test/Conversion/XTenNNToTorch/conv_transpose_2d.mlir:99]: Added and done.

File: lib/Conversion/XTenNNToTorch.cpp (hunk 1 above)
Comment (ehsan-toosi) [reply lib/Conversion/XTenNNToTorch.cpp]: Haha that was supposed to be a TODO that I forgot to check. You guys did. It's implemented now.

**Review by roberteg16 — APPROVED** [review]
LGTM, thanks!

**Review by ttjost — APPROVED** [review]
